### PR TITLE
Instrument authenticators.

### PR DESCRIPTION
**Pull request by immutableT**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Since a typical cluster today uses multiple types of authentications, it is important to classify authentication related metrics based on the type (ex. x509 vs. bearer token).

**Does this PR introduce a user-facing change?**:
```release-note
This PR adds a new label - "authenticator" for the following three metrics:
authenticated_user_requests
authentication_attempts
authentication_duration_seconds
```

### CODE REVIEW

**Comment by immutableT** [comment]
/assign @timstclair 
@timstclair PTAL. Did I capture the idea you suggested on #85113?
 I know that I still need to implement the logic for the token and password, though it should be fairly similar.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 1]
```diff
@@ -29,6 +29,44 @@ type Token interface {
...
+
+// AuthenticateRequest wraps AuthenticateRequest of the delegate Request.
+func (i *InstrumentedRequest) AuthenticateRequest(req *http.Request) (*Response, bool, error) {
+	resp, ok, err := i.delegate.AuthenticateRequest(req)
+	if err != nil {
+		err = InstrumentedError{Name: i.Name, delegate: err}
+	}
+
+	if resp != nil {
+		resp.AuthenticatorName = i.Name
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Is it possible that another authenticator has already set the authenticator name on the response?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 2]
```diff
@@ -29,6 +29,44 @@ type Token interface {
...
+}
+
+// AuthenticateRequest wraps AuthenticateRequest of the delegate Request.
+func (i *InstrumentedRequest) AuthenticateRequest(req *http.Request) (*Response, bool, error) {
+	resp, ok, err := i.delegate.AuthenticateRequest(req)
+	if err != nil {
+		err = InstrumentedError{Name: i.Name, delegate: err}
+	}
+
+	if resp != nil {
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Should this only set the AuthenticatorName if the request is authenticated?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go [hunk 3]
```diff
@@ -65,15 +65,15 @@ func TestMetrics(t *testing.T) {
 			want: `
         # HELP authentication_attempts [ALPHA] Counter of authenticated attempts.
         # TYPE authentication_attempts counter
-        authentication_attempts{result="error"} 1
+        authentication_attempts{authenticator="unknown", result="error"} 1
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go]: Can you update these tests to use a named authenticator?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 4]
```diff
@@ -29,6 +29,44 @@ type Token interface {
...
+	}
+
+	return resp, ok, err
+}
+
+// InstrumentedError implementation of the Error interface, and is intended for use by the
+// InstrumentedRequest.
+type InstrumentedError struct {
+	// Name identifies the delegate authenticator that produced the error.
+	Name     string
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: nit: prefer `AuthenticatorName`

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 5]
```diff
@@ -29,6 +29,44 @@ type Token interface {
...
+	delegate error
+}
+
+// Error implementation of the Error interface.
+func (e InstrumentedError) Error() string {
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: nit: if you embed the delegate, you don't need this function.

**Review by tallclair — COMMENTED** [review]
Yep, this looks good.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 1 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: I think it is (ex. union).
Added a check.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go (hunk 3 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 4 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 5 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 2 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Good question: I would imagine we want to instrument requests that failed authentication as well. So, not checking for "ok" seems reasonable. WDYT?

Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: If it failed authentication, that really means it failed authentication by all authenticators, right? Or would this be for the case where it provided a specific type of credential, but it was invalid?

Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: To elaborate, if I make a request without any credentials, I don't want the last (or first) authenticator in the chain to be attributed the failure, otherwise that will be misleading.

**Review by tallclair — COMMENTED** [review]
Looks good, let me know when the remaining pieces are completed.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 2 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: As written, the failure (in the scenario where all authenticators in a union fail) will be attributed to the "unknown" authenticator, which is not what we want (see ""union intersection - all authenticators fail" test-case). 
This is due to [AuthenticateRequest](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/authentication/request/union/union.go#L56) being delegated to the "union" authenticator which is not instrumented.

One option is to instrument the "union" authenticators (request and token).

An alternative would be to instrument InstrumentedAuthenticator, InstrumentedToken and InstrumentedPassword, which seems logical. 
WDYT?

Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Ended-up instrumenting InstrumentedAuthenticator, InstrumentedToken and InstrumentedPassword - I believe we now get the expected behaviour.

File: pkg/kubeapiserver/authenticator/config.go [hunk 6]
```diff
@@ -128,7 +153,7 @@ func (config Config) New() (authenticator.Request, *spec.SecurityDefinitions, er
 	// X509 methods
 	if config.ClientCAContentProvider != nil {
 		certAuth := x509.NewDynamic(config.ClientCAContentProvider.VerifyOptions, x509.CommonNameUserConversion)
-		authenticators = append(authenticators, certAuth)
+		authenticators = append(authenticators, authenticator.WrapInInstrumentedRequest(certAuth, "x509"))
```
Comment (tallclair) [inline pkg/kubeapiserver/authenticator/config.go]: how come you didn't use a constant for x509?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 7]
```diff
@@ -29,6 +30,71 @@ type Token interface {
...
+// AuthenticateRequest wraps AuthenticateRequest of the delegate Request.
+func (i *InstrumentedRequest) AuthenticateRequest(req *http.Request) (*Response, bool, error) {
+	authenticationStart := time.Now()
+	resp, ok, err := i.delegate.AuthenticateRequest(req)
+	defer recordAuthMetrics(req.Context(), resp, ok, err, authenticationStart)
+	if err != nil {
+		err = InstrumentedError{AuthenticatorName: i.AuthenticatorName, error: err}
+	}
+
+	// TODO(immutableT) I don't believe we could ever receive a response with AuthenticatorName already set.
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: I don't think this should be a todo - Just make it a regular comment, and log something if the assumption is violated.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics_test.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+				RequestFunc(
+					func(_ *http.Request) (*Response, bool, error) {
+						return &Response{
+								User:      &user.DefaultInfo{Name: "admin"},
+								Audiences: Audiences{"audience-x"},
+							},
+							true,
+							nil
+					}),
+				"instrumented-auth"),
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics_test.go]: Can you add a couple tests using the union authenticator?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go [hunk 9]
```diff
@@ -17,131 +17,143 @@ limitations under the License.
 package filters
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go]: Shouldn't this be moved along with metrics.go? It looks like you added a second metrics_test.go, so maybe merge the two?

**Review by tallclair — COMMENTED** [review]
Looks good, just a couple little cleanups.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go (hunk 9 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go]: I had to split-out tests that use the "union" package to avoid circular dependency between authenticator and filter packages. 
An alternative could be to create a separate testing package.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics_test.go (hunk 8 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics_test.go]: Union tests are in the filter/metrics_test, see comment below on why I split them out.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 7 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Done.

File: pkg/kubeapiserver/authenticator/config.go (hunk 6 above)
Comment (immutableT) [reply pkg/kubeapiserver/authenticator/config.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go (hunk 9 above)
Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go]: I see, in that case I'd prefer one of these options:
Create a `staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics` package
**or**
Move this test to the union package

Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics_test.go]: Moved into the union package.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 10]
```diff
@@ -77,4 +171,6 @@ type Response struct {
 	Audiences Audiences
 	// User is the UserInfo associated with the authentication context.
 	User user.Info
+	// AuthenticatorName identifies the authenticator that produced the response.
+	AuthenticatorName string
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: I see no reason to expand `Response`

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 11]
```diff
@@ -29,6 +30,71 @@ type Token interface {
 	AuthenticateToken(ctx context.Context, token string) (*Response, bool, error)
 }
 
+// WrapInInstrumentedToken wraps the input Token in the InstrumentedRequest.
+func WrapInInstrumentedToken(token Token, authenticatorName string) *InstrumentedToken {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: These wrappers should be in a different package (or at the very least a different file).

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 12]
```diff
@@ -29,6 +30,71 @@ type Token interface {
...
+	return &InstrumentedToken{AuthenticatorName: authenticatorName, delegate: token}
+}
+
+// InstrumentedToken wraps Token and allows instrumentation to identify authenticators via the AuthenticatorName field.
+type InstrumentedToken struct {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Does this type need to be exported?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 13]
```diff
@@ -29,6 +30,71 @@ type Token interface {
...
+	authenticationStart := time.Now()
+	resp, ok, err := i.delegate.AuthenticateToken(ctx, token)
+	defer recordAuthMetrics(ctx, resp, ok, err, authenticationStart)
+	if err != nil {
+		err = InstrumentedError{AuthenticatorName: i.AuthenticatorName, error: err}
+	}
+
+	if resp != nil && resp.AuthenticatorName == "" {
+		resp.AuthenticatorName = i.AuthenticatorName
+	}
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Why?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 14]
```diff
@@ -29,6 +30,71 @@ type Token interface {
...
+	return resp, ok, err
+}
+
+// InstrumentedError implementation of the Error interface, and is intended for use by the
+// InstrumentedRequest.
+type InstrumentedError struct {
+	// AuthenticatorName identifies the delegate authenticator that produced the error.
+	AuthenticatorName string
+	error
+}
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Not sure why we need this - just pass in `i.AuthenticatorName` to `recordAuthMetrics`.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go [hunk 15]
```diff
@@ -75,21 +76,43 @@ func init() {
...
+
+	if resp != nil && resp.AuthenticatorName != "" {
+		authenticatorName = resp.AuthenticatorName
+	}
 
 	switch {
 	case err != nil || (resp != nil && !audiencesAreAcceptable(apiAudiences, resp.Audiences)):
 		resultLabel = errorLabel
-	case !ok:
+		if errors.As(err, &instrumentedError) && instrumentedError.AuthenticatorName != "" {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: Pretty sure this is passing in a `**InstrumentedError`?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go [hunk 16]
```diff
@@ -75,21 +76,43 @@ func init() {
 	legacyregistry.MustRegister(authenticationLatency)
 }
 
-func recordAuthMetrics(resp *authenticator.Response, ok bool, err error, apiAudiences authenticator.Audiences, authStart time.Time) {
-	var resultLabel string
+// ResetMetrics resets all auth-related metrics - intended for use in unit tests only.
+// Since prometheus' gatherer is global, other tests may have updated metrics already, so
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: 😞

**Review by enj — CHANGES_REQUESTED** [review]
I think this is too late for v1.18.  Code that touches the entire authentication stack needs time to soak in CI.  I suggest merging this as early as possible in v1.19 to allow for ample time for us to notice regressions.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 14 above)
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Maybe for the union case?  Even so, I would expect the instrumented authenticators to tell you their name directly, not encode it by wrapping errors.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 11 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Moved into metrics.go file (in the same package).

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 13 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Removed.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 14 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Removed.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go (hunk 15 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: Removed the InstrumentedError type.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 10 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: Right. Reverted the change.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 12 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: No (same for the other instrumented types) - unexported.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go (hunk 16 above)
Comment (brancz) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: One of our future initiatives is to remove the use of the global metrics registry. You can already use a registry that you pass through your code and register metrics onto it. For now you would just pass the global registry but for tests you could already use a local one.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+const (
+	successLabel = "success"
+	failureLabel = "failure"
+	errorLabel   = "error"
+)
+
+var (
+	authenticatedUserCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Name:           "authenticated_user_requests",
```
Comment (brancz) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: This needs the _total suffix.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+			Help:           "Counter of authenticated requests broken out by username.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"username", "authenticator"},
+	)
+
+	authenticatedAttemptsCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Name:           "authentication_attempts",
```
Comment (brancz) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: Same as above.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go (hunk 16 above)
Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: You can create a test registry by [NewFakeKubeRegistry](https://github.com/kubernetes/kubernetes/blob/5708511499fe500ae3b4bbd40204cef382f652e8/staging/src/k8s.io/component-base/metrics/testutil/testutil.go#L57). If it can't fully fit your case, please feel free let me know.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+	// authenticatorName identifies the delegate authenticator.
+	authenticatorName string
+	delegate          Request
+}
+
+// AuthenticateRequest wraps AuthenticateRequest of the delegate Request.
+func (i *instrumentedRequest) AuthenticateRequest(req *http.Request) (*Response, bool, error) {
+	authenticationStart := time.Now()
+	resp, ok, err := i.delegate.AuthenticateRequest(req)
+	recordAuthMetrics(req.Context(), resp, i.authenticatorName, ok, err, authenticationStart)
```
Comment (tallclair) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: Is this going to result in every authenticator logging a deny when a user is rejected?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go (hunk 17 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: This is an exiting metric, would not this be a breaking change?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go (hunk 19 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: Yes. Is not it what we want?

Comment (tallclair) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: I don't think so?

If a user is authenticated, we want to know which authenticator authenticated them. If the authentication is rejected, I'm not sure how useful it is to know which authenticator rejected them. Unless we could distinguish between "had the right type of credentials, but they were invalid" from "didn't have the credentials".

Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: I assume that the case of "had the right type of credentials, but they were invalid" results in !ok, while the case of "didn't have the credentials" results in an error. If that is true, then we do end-up with different labels on the metric: failure vs. error.

I think I need to add more tests to confirm this.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go (hunk 16 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: Thank for the tip on NewFakeKubeRegistry.
Using an isolated fake now.

**Comment by liggitt** [comment]
I agree with @enj that this is a big change to be landing this late.

/milestone v1.19

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go (hunk 19 above)
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/metrics.go]: I reviewed several authenticators, and it appears that we don't return errors when an incompatible request is received by an authenticator - this makes sense in the context of a union. Instead, we return `nil, false nil` (nil for both: response and error) when the request is not compatible (not what the authenticator expected).
Therefore, I added a condition that will **not** record a metric in such cases. In other words, receiving an incompatible request does not count as processing a request.

However, this implies that when all of the authenticators within a union are incompatible with a request, (as written) no metric will be recorded - this feels wrong since such events are  interesting from monitoring perspective.

I feel that the union package should return an error when none of the authenticators considered the request to be compatible. Though this feels like a significant change in the API. WDYT?

**Comment by enj** [comment]
@immutableT, in the sig-auth meeting today @tallclair mentioned that some of the data this PR is trying to emit should be done via the audit stack through the changes that #89305 enables.  The name of the authenticator seems like a good candidate for an annotation.

**Comment by tallclair** [comment]
@immutableT - ping on this PR. Can we get this into v1.19?

**Comment by immutableT** [comment]
@eng and @tallclair, I would like to confirm my understanding of how this PR should leverage Audit Annotations (#89305):

I assume that we could use Audit Annotations as a way for requests to identify themselves by injecting their names into the context, correct? This is in contrast to the approach that this PR takes of wrapping requests in the InstrumentedRequest, and the process of wrapping attaching their names.

**Comment by tallclair** [comment]
IMO we want both, instrumentation and audit. I think we should get this PR submitted for instrumentation, and follow up with a separate PR to add audit annotations.

**Comment by immutableT** [comment]
I see. So, the ping was just to re-base, right?

**Comment by tallclair** [comment]
I'm not sure. @enj did you have a specific ask for this PR, or was that just an FYI?

**Comment by enj** [comment]
Oh I may have misunderstood what @tallclair was asking for in the sig-auth meeting.  I thought the intention was to solely use the audit annotations approach (presumably because they contain all information the metrics would contain and more without any concerns around cardinality).

Do we feel that there is enough value in having the same-ish information in two places?  Just wondering if the increase in API surface makes sense.

**Comment by tallclair** [comment]
IMO there's still value to the metrics, even if the audit logs contain more detailed information. We could technically scan the audit logs to generate metrics, but I think this will be a common enough need that it makes sense to build it in.

**Comment by enj** [comment]
> IMO there's still value to the metrics, even if the audit logs contain more detailed information. We could technically scan the audit logs to generate metrics, but I think this will be a common enough need that it makes sense to build it in.

SGTM.  I will try to review this soon so we can merge it early in the v1.19 release cycle.

**Comment by jtslear** [comment]
Hello, @jtslear here from Bug Triage! This PR has not been updated for a long time, so I'd like to check on the status. The code freeze is starting June 25th and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

@enj based on the last comment we wanted to get this in as soon in this current cycle, as the PR is tagged for 1.19, is it still planned for this release?

**Comment by jtslear** [comment]
@kubernetes/sig-api-machinery-misc @kubernetes/sig-auth-pr-reviews @kubernetes/sig-instrumentation-pr-reviews 

We haven't seen any further progress made in the last 30 days.  Is this PR still valid for v1.19?

**Comment by enj** [comment]
> Is this PR still valid for v1.19?

Yes, this is on my list to review this week.

File: staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics_test.go [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+					"instrumented-auth-02"),
+			),
+			want: `
+       # HELP authenticated_user_requests [ALPHA] Counter of authenticated requests broken out by username.
+        # TYPE authenticated_user_requests counter
+        authenticated_user_requests{authenticator="instrumented-auth-02",username="admin"} 1
+        # HELP authentication_attempts [ALPHA] Counter of authenticated attempts.
+        # TYPE authentication_attempts counter
+        authentication_attempts{authenticator="instrumented-auth-01",result="failure"} 1
+        authentication_attempts{authenticator="instrumented-auth-02",result="success"} 1
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics_test.go:253]: I feel like this will just be super noisy for every form of auth that is early in the chain.

I would like to see how this looks for a real series of authenticators:

1. request header
1. x509
1. legacy SA
1. new SA
1. oidc (might be hard to set up)
1. token webhook (might be hard to set up)
1. anonymous

You would then pass an credential meant for each authenticator and see what the overall metrics look like.

**Review by enj — CHANGES_REQUESTED** [review]
This looks pretty close to me.

File: staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics.go [hunk 21]
```diff
@@ -75,21 +76,33 @@ func init() {
...
 	default:
 		resultLabel = successLabel
-		authenticatedUserCounter.WithLabelValues(compressUsername(resp.User.GetName())).Inc()
+		authenticatedUserCounter.WithLabelValues(compressUsername(resp.User.GetName()), authenticatorName).Inc()
 	}
 
-	authenticatedAttemptsCounter.WithLabelValues(resultLabel).Inc()
-	authenticationLatency.WithLabelValues(resultLabel).Observe(time.Since(authStart).Seconds())
+	authenticatedAttemptsCounter.WithLabelValues(resultLabel, authenticatorName).Inc()
+	authenticationLatency.WithLabelValues(resultLabel, authenticatorName).Observe(time.Since(authStart).Seconds())
```
[the remark this answers is not in the document]
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics.go:105]: I believe that the noise is removed by this check:

```
if resp == nil && err == nil {
		// When both of these are nil, we assume that the request was incompatible with the authenticator.
		return
}
```

Right?

File: staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics.go [hunk 22]
```diff
@@ -75,21 +76,33 @@ func init() {
...
-func recordAuthMetrics(resp *authenticator.Response, ok bool, err error, apiAudiences authenticator.Audiences, authStart time.Time) {
+func recordAuthMetrics(ctx context.Context, resp *authenticator.Response, authenticatorName string, authSuccess bool, err error, authStart time.Time) {
 	var resultLabel string
 
+	if resp == nil && err == nil {
+		// When both of these are nil, whe assume that the request was incompatible with the authenticator.
+		return
+	}
+
+	if authenticatorName == "" {
```
[the remark this answers is not in the document]
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics.go:87]: Done.

File: staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics_test.go [hunk 23]
```diff
@@ -0,0 +1,327 @@
...
+			request: WrapRequest(
+				authenticator.RequestFunc(
+					func(_ *http.Request) (*authenticator.Response, bool, error) {
+						return nil, false, errors.New("some error")
+					}),
+				"instrumented-auth"),
+			want: `
+		    # HELP authentication_attempts [ALPHA] Counter of authenticated attempts.
+		    # TYPE authentication_attempts counter
+		    authentication_attempts{authenticator="instrumented-auth", result="error",} 1
```
[the remark this answers is not in the document]
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics_test.go:73]: I see your point.

How would we define an alert based on the error rate per authenticator? 
It does not seem to be possible to do so since we can't differentiate these two scenarios:
1. webhook auth is having a problem
2. the token was intended to another token authenticator higher up the stack

[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics_test.go:73]: I think we might have to make this stateful by keeping track of all metrics that we may want to assert.  On success, we would only report the last one (i.e. the successful authentication with authenticator named `foo`).  On failure, we would report failure metrics for all authenticators where `err != nil`.  On the `nil, false, nil` case, we report no metrics related to authenticator name.

WDYT?

[the remark this answers is not in the document]
Comment (immutableT) [reply staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics_test.go:73]: I believe there five use-cases:
1. An authenticator receives an incompatible credential. In this case, such an authenticator should return `nil, false, nil` and such authentication attempts should be ignored in the `recordMetrics` since this is expected.
2. An authenticator receives a compatible credential, but the authentication fails (ex. incorrect password). In this case, `nil, false, nil` is **also** returned. This case is interesting from the monitoring point of view since it may indicate a security incident - somebody is trying to brute-force a password, and I think we should be recording this.
3. The supplied credential is compatible with the authenticator, but an error occurred during validation (ex. network failure), in this case we return, nil, false, someError`. We certainly, want to record this fact.
4. Happy path - compatible credentials and successful authentication.
5. None of the authenticators were able to authenticate a request/token.

Therefore
1. As you suggested, to distinguish between 1 and 2 we need to know the final result produced by the union authenticator. If the final outcome is a success, we could ignore case 2. 
2. For use-case 5, I think we need to instrument our union authenticators - they could have their own instrumentation to track requests that failed the chain of authenticators.
3. The current state of this PR properly handles cases 3 and 4.

Actually, by instrumenting union authenticators (to emit a metric when none of the authenticators in the chain were successful) we can safely ignore the issue of not being able to distinguish cases 1 and 2. In other words, if there a brute-force attack in progress, this would be captured by the union authenticator. 

> On the nil, false, nil case, we report no metrics related to authenticator name.
   I believe this is already what this PR does.
```
if resp == nil && err == nil {
  // When both of these are nil, we assume that the request was incompatible with the authenticator.
  return
}
```

In summary, I think the current logic in this PR is correct, we just need to instrument the union authenticators. WDYT?

**Comment by jtslear** [comment]
Due to reaching the code freeze deadline, I'm going to bump this to the next release.

/milestone v1.20

**Comment by mkorbi** [comment]
Hello @immutableT 
This PR has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting 12.11.2020 (about 3-4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
As the PR is tagged for 1.20, is it still planned for this release?

**Comment by tallclair** [comment]
This PR needs someone else to adopt it (for v1.21), as @immutableT is no longer working on Kubernetes.

**Comment by vinayakankugoyal** [comment]
> This PR needs someone else to adopt it (for v1.21), as @immutableT is no longer working on Kubernetes.

I'd like to adopt this.

**Comment by ehashman** [comment]
> This PR needs someone else to adopt it (for v1.21), as @immutableT is no longer working on Kubernetes.

I'll close this then, and hopefully someone else can pick this up! @vinayakankugoyal I suggest you open a new PR.

/close
